File: runtime/controller/authority.ts
import { PublicKey } from "@solomon-labs/publickey";

import { AUTHORITY_ROLE } from "../../clients/ts/src/constants.js";
import type { ControllerAuthority, ControllerAuthorityRole, ControllerConfiguration } from "./types.js";

const AUTHORITY_SEED = new TextEncoder().encode("chancery-authority");
const FAUCET_AUTHORITY_SEED = new TextEncoder().encode("faucet-authority");

export const CONTROLLER_ROLES: readonly ControllerAuthorityRole[] = [
    AUTHORITY_ROLE.GOVERNANCE,
    AUTHORITY_ROLE.OPS,
    AUTHORITY_ROLE.EMERGENCY,
    AUTHORITY_ROLE.ENFORCEMENT,
    AUTHORITY_ROLE.INSURANCE_ADMIN,
];

export async function controllerConfiguration(programId: PublicKey): Promise<ControllerConfiguration> {
    if (programId.toBase58() === "11111111111111111111111111111111") {
        throw new Error("controller program identity is unstamped; run yarn identity");
    }
    const authorities = await Promise.all(CONTROLLER_ROLES.map(async (role): Promise<ControllerAuthority> => {
        const [publicKey] = await PublicKey.findProgramAddress([AUTHORITY_SEED, new Uint8Array([role])], programId);
        return { role, publicKey };
    }));
    return { programId, authorities };
}

export function controllerAuthority(configuration: ControllerConfiguration, role: ControllerAuthorityRole): PublicKey {
    const authority = configuration.authorities.find((entry) => entry.role === role);
    if (authority === undefined) throw new Error("controller has no authority for role " + role);
    return authority.publicKey;
}

export async function faucetAuthority(programId: PublicKey): Promise<PublicKey> {
    const [authority] = await PublicKey.findProgramAddress([FAUCET_AUTHORITY_SEED], programId);
    return authority;
}

File: programs/faucet/src/authority.rs
use chancery::constants::authority_role;
use solana_program_error::ProgramError;
use solana_pubkey::Pubkey;

use crate::constants::CHANCERY_AUTHORITY_SEED;

pub const ROLES: [u8; 5] = [
    authority_role::GOVERNANCE,
    authority_role::OPS,
    authority_role::EMERGENCY,
    authority_role::ENFORCEMENT,
    authority_role::INSURANCE_ADMIN,
];
pub const ALL_ROLES_MASK: u8 = 0x1f;

pub fn validate_role_mask(mask: u8) -> Result<(), ProgramError> {
    if mask == 0 || mask & !ALL_ROLES_MASK != 0 {
        return Err(ProgramError::InvalidInstructionData);
    }
    Ok(())
}

pub fn authority_address(role: u8, program_id: &Pubkey) -> Result<(Pubkey, u8), ProgramError> {
    if !ROLES.contains(&role) {
        return Err(ProgramError::InvalidInstructionData);
    }
    Ok(Pubkey::find_program_address(&[CHANCERY_AUTHORITY_SEED, &[role]], program_id))
}
